### PR TITLE
Dispatch errors from the primary response to deferred responses

### LINKED ISSUE #1818: @defer: error related to a field in a deferred fragment appears in chunk for non-deferred fragment
**Describe the bug**
Using the graphql [test suite](https://github.com/graphql/graphql-js/blob/a24a9f35b876bdd0d5050eca34d3020bd0db9a29/src/execution/__tests__/defer-test.ts#L424) as an inspiration, I am selecting a field that resolves to an error, in a deferred fragment:

```graphql
query HandlesErrorsThrownInDeferredFragmentsQuery {
  computer(id: "Computer1") {
    id
    ...ComputerErrorField @defer
  }
}

fragment ComputerErrorField on Computer {
  errorField
}
```

I am receiving the error in the first chunk (along with `id`) and no errors in the second chunk, whereas I would have expected the opposite (error in the 2nd chunk along `errorField`), as graphql-js apparently does.

Unsure if this is the spec really mandates this, it may be just a different and acceptable implementation, if so feel free to close!

```
--graphql
content-type: application/json

{"data":{"computer":{"id":"Computer1"}},"errors":[{"message":"Subgraph errors redacted"}],"hasNext":true}
--graphql
content-type: application/json

{"hasNext":false,"incremental":[{"data":{"errorField":null},"path":["computer"]}]}
--graphql--
```

**To Reproduce**

```
git clone git@github.com:apollographql/apollo-kotlin.git
cd apollo-kotlin && git checkout defer-with-router-tests
(cd tests/defer/router/subgraphs/computers && npm install && APOLLO_PORT=4001 npm start)&
path/to/router --supergraph tests/defer/router/simple-supergraph.graphqls &
curl --request POST     --header 'content-type: application/json'     --url http://localhost:4000     --data '{"query":"query HandlesErrorsThrownInDeferredFragmentsQuery {  computer(id: \"Computer1\") {    id    ...ComputerErrorField @defer  } }  fragment ComputerErrorField on Computer {  errorField }"}' --header 'Accept: multipart/mixed; deferSpec=20220824'
```

### LINKED ISSUE #2185: When processing deferred errors, they must be properly dispatched to each array element
See [#2183 for more details.](https://github.com/apollographql/router/pull/2183#discussion_r1036219633)

### PR DESCRIPTION
**Pull request by Geal**
Fix #1818 
Fix #2185
Blocked by #2109 (waiting for the router-bridge update)

When errors are generated during the primary execution, some of them can be affected to
deferred responses.

To handle that, we need to:
- transmit errors from the primary query to deferred node execution along with the primary fetches
- be able to check if an error path belongs to a deferred query

Since the error path may belong to a part of the response that was nullified, we need to follow the error path through the primary or deferred queries

### CODE REVIEW

File: apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__deferred_fragment_bounds_nullability-2.snap [hunk 1]
```diff
@@ -23,24 +23,6 @@ expression: stream.next_response().await.unwrap()
               "suborga",
               0
             ]
-          },
```
Comment (Geal) [inline apollo-router/src/services/snapshots/apollo_router__services__supergraph_service__tests__deferred_fragment_bounds_nullability-2.snap:26]: this is removing the redundant parts of the snapshot, now that error dispatching is fixed

File: NEXT_CHANGELOG.md [hunk 2]
```diff
@@ -117,6 +117,12 @@ By [@Geal](https://github.com/geal) in https://github.com/apollographql/router/p
 When we drop Telemetry we spawn a thread to perform the global opentelemetry trace provider shutdown. The documentation of this function indicates that "This will invoke the shutdown method on all span processors. span processors should export remaining spans before return". We should give that process some time to complete (5 seconds currently) before returning from the `drop`. This will provide more opportunity for spans to be exported.
 
 By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/2191
+### Dispatch errors from the primary response to deferred responses ([Issue #1818](https://github.com/apollographql/router/issues/1818), [Issue #2185](https://github.com/apollographql/router/issues/2185))
+
+When errors are generated during the primary execution, some of them can be affected to
```
[the remark this answers is not in the document]
Comment (Geal) [reply NEXT_CHANGELOG.md:130]: ah right, in french "affecter" can also have the meaning to "assign" :sweat_smile: . Here I mean that some of the errors generated in the primary execution should actullay be sent as part of the deferred response

File: apollo-router/src/services/execution_service.rs [hunk 3]
```diff
@@ -166,25 +174,69 @@ where
...
+                                        .collect();
+
                                     // an empty response should not be sent
                                     // still, if there's an error or extension to show, we should
                                     // send it
                                     if !data.is_null()
                                         || !errors.is_empty()
-                                        || !response.extensions.is_empty()
+                                        || !extensions.is_empty()
+
```
Comment (o0Ignition0o) [inline apollo-router/src/services/execution_service.rs:230]:
```suggestion
```

File: apollo-router/src/services/execution_service.rs [hunk 4]
```diff
@@ -166,25 +174,69 @@ where
...
                                     {
                                         Some(
                                             IncrementalResponse::builder()
                                                 .and_label(response.label.clone())
                                                 .data(data)
                                                 .path(path)
                                                 .errors(errors)
-                                                .extensions(response.extensions.clone())
+                                                .extensions(extensions)
+
```
Comment (o0Ignition0o) [inline apollo-router/src/services/execution_service.rs:239]:
```suggestion
```

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**